File: src/augmenters/traffic_light.py
from sortedcontainers import SortedList  # type: ignore
from typing_extensions import override

from ..drawer_utils import CapacityEvent, Interface, UserInterface, dtPoint
from .base_augmenter import TrafficAugmenter


class TrafficLight(TrafficAugmenter):
    """Specialization of augmenter for traffic lights. Traffic light augmenters cause capacity to
    drop to 0 for a specific period of time, followed by the release of that limitation.
    This occurs is a configurable cyclical manner.
    """

    id: int = 0

    def __init__(
        self,
        pos: float,
        cycles: tuple[float, ...],
        blocking_states: tuple[bool, ...],
        init_state: int = 0,
        delay: float = 0,
    ):
        """Traffic light constructor.

        Args:
            pos (float): the position at which the traffic light is located (meters)
            cycles (tuple[float, ...]): the times of each traffic light cycle
            blocking_states (tuple[bool, ...]): whether or not each cycle is blocking or not
            init_state (int, optional): initial state of the traffic light. Defaults to 0.

        Raises:
            ValueError: the state must be a valid index of cycle
            ValueError: each cycle of the traffic light must be defined to be blocking or not
        """
        self.pos = pos  # where the traffic light is
        self.cycles = cycles  # list of the delays of each of the cycles

        # initial state -- an index for self.cycles (will wrap around)
        self.init_state = init_state
        self.blocking_states = (
            blocking_states  # list of whether each corresponding cycle is blocking or not
        )

        self.delay = delay

        if not delay >= 0:
            raise ValueError("The delay must be positive")

        if not (self.init_state < len(self.cycles) and self.init_state >= 0):
            raise ValueError("The provided initial state is invalid.")

        if len(blocking_states) != len(cycles):
            raise ValueError("The lengths of the blocking states and cycles don't match")

        # generate an unique id
        self.id = TrafficLight.id
        TrafficLight.id += 1

    @override
    def init(self, simulation_time: float, events: SortedList, interfaces: list[Interface]):
        time = self.delay
        state = self.init_state

        # continue adding capacity events to the event queue until we are out of time
        while time <= simulation_time:
            if self.blocking_states[state]:
                start = dtPoint(time, self.pos)
                end = dtPoint(time + self.cycles[state], self.pos)

                cur = UserInterface(start, 0, self, lower_bound=start, upper_bound=end)
                interfaces.append(cur)

                start_event = CapacityEvent(start, cur, posterior_capacity=0)
                events.add(start_event)

                end_event = CapacityEvent(end, cur, prior_capacity=0)
                events.add(end_event)

            time += self.cycles[state]
            state = (state + 1) % len(self.cycles)

File: src/augmenters/traffic_light.ts
import { CapacityEvent, UserInterface, dtPoint } from "../drawer_utils";
import { ShockwaveDrawer } from "../shockwave_drawer";

import { CapacityBottleneck } from "./base_augmenter";


export class TrafficLight extends CapacityBottleneck {
  private pos: number;
  private cycles: number[];
  private blockingStates: number[];
  private initState: number;
  private delay: number;

  constructor(
    pos: number,
    cycles: number[],
    blockingStates: number[],
    initState = 0,
    delay = 0
  ) {
    super(0);

    this.pos = pos;
    this.cycles = cycles;
    this.blockingStates = blockingStates;
    this.initState = initState;
    this.delay = delay;

    if (!(this.delay >= 0)) {
      throw new RangeError("Delay must be positive");
    }

    if (!(this.initState < this.cycles.length && this.initState >= 0)) {
      throw new RangeError("Initial state provided is invalid");
    }

    if (this.blockingStates.length !== this.cycles.length) {
      throw new RangeError("Length of blocking state/cycle arrays do not match");
    }
  }

  public init(drawer: ShockwaveDrawer) {
    let time = this.delay;
    let state = this.initState;

    while (time <= drawer.getSimulationTime()) {
      if (this.blockingStates[state]) {
        const start = new dtPoint(time, this.pos);
        const end = new dtPoint(time + this.cycles[state], this.pos);

        const cur = new UserInterface(start, 0, this, start, end);
        drawer.addUserInterface(cur);

        const startEvent = new CapacityEvent(start, cur, undefined, 0);
        drawer.addCapacityEvent(startEvent);

        const endEvent = new CapacityEvent(end, cur, 0);
        drawer.addCapacityEvent(endEvent);
      }

      time += this.cycles[state];
      state = (state + 1) % this.cycles.length;
    }
  }
}